source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,409 @@
+    {
+        auto pk = Scalar(secret_key);
+        auto pk_point = pk.toPoint();
+        auto signature = sign(pk, pk_point, signature_noise.V,
+            signature_noise.v, hash[]);
+
+        return signature;
+    }
+
+    /// load random seed from file
+    protected void loadNodeEnrollmentData ()
+    {
+        if (this.node_enroll_data_path.exists())
+        {
+            auto seed_file = File(this.node_enroll_data_path, "r");
+            scope DeserializeDg dg = (size) @safe
+            {
+                ubyte[] res;
+                res.length = size;
+                seed_file.rawRead(res);
+                return res;
+            };
+
+            this.random_seed_src.deserialize(dg);
+            this.signature_noise.deserialize(dg);
+        }
+    }
+
+    /// save random seed to file
+    protected void saveNodeEnrollmentData ()
+    {
+        auto seed_file = File(this.node_enroll_data_path, "w");
+        this.random_seed_src.serialize((scope bytes) => seed_file.rawWrite(bytes));
+        this.signature_noise.serialize((scope bytes) => seed_file.rawWrite(bytes));
+    }
+}
+
+/// tests for member functions of EnrollmentManager
+unittest
+{
+    import agora.common.Amount;
+    import agora.consensus.data.Transaction;
+    import agora.consensus.data.UTXOSet;
+    import std.format;
+
+    KeyPair[] key_pairs = [KeyPair.random];
+
+    class UnitEnrollMan : EnrollmentManager
+    {
+        this () { super(":memory:", "", key_pairs[0]); }
+        protected override void loadNodeEnrollmentData () {}
+        protected override void saveNodeEnrollmentData () {}
+    }
+
+    // create the first transaction
+    Transaction firstTx = Transaction(
+        TxType.Freeze,
+        [Input(Hash.init, 0)],
+        [Output(Amount(100_000), key_pairs[0].address)]
+    );
+    Hash firstHash = hashFull(firstTx);
+
+    // create the second transaction
+    Transaction secondTx = Transaction(
+        TxType.Freeze,
+        [Input(Hash.init, 0)],
+        [Output(Amount(50_000), key_pairs[0].address)]
+    );
+    Hash secondHash = hashFull(secondTx);
+
+    // create an EnrollmentManager object
+    auto utxo_set = new UTXOSet(":memory:");
+    scope (exit) utxo_set.shutdown();
+    auto man = new UnitEnrollMan();
+
+    // create and add the first Enrollment object
+    auto utxo_hash = utxo_set.getHash(firstHash, 0);
+    auto enroll = man.createEnrollment(utxo_hash);
+    assert(man.hasEnrollment(utxo_hash) == false);
+    man.addEnrollment(enroll);
+    assert(man.getEnrollmentLength() == 1);
+    assert(man.hasEnrollment(utxo_hash) == true);
+
+    // create and add the second Enrollment object
+    utxo_hash = utxo_set.getHash(secondHash, 0);
+    enroll = man.createEnrollment(utxo_hash);
+    man.addEnrollment(enroll);
+    assert(man.getEnrollmentLength() == 2);
+
+    // remove an Enrollment object
+    man.removeEnrollment(utxo_hash);
+    assert(man.getEnrollmentLength() == 1);
+
+    man.shutdown();
```

Comment:
This calls the shutdown a second time after `scope (exit) utxo_set.shutdown();` already did it. So you can just remove `man.shutdown()` here.

Reply:
`man.shutdown()` must be called. So I will add the `scope(exit) man.shutdown() ` below the `auto man = new UnitEnrollMan()` code